src/xdg.rs

```diff
@@ -221,42 +208,125 @@ impl Icon {
     }
 }
 
+/// Guessed types of images.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Bucket {
+    /// A bitmap image of a known square size.
+    Exact(u32),
+
+    /// A bitmap image of an unknown size.
+    Unknown,
+
+    /// A vector image.
+    Scalable,
+
+    /// A monochrome vector image.
+    Symbolic,
+}
+
 /// Simple loader for app icons.
 #[derive(Debug)]
 struct IconLoader {
-    icons: HashMap<String, PathBuf>,
+    icons: HashMap<String, HashMap<Bucket, PathBuf>>,
 }
 
 impl IconLoader {
     /// Initialize the icon loader.
     ///
     /// This will check all paths for available icons and store them for cheap
     /// lookup.
-    fn new(data_dirs: &[PathBuf]) -> Self {
-        let mut icons = HashMap::new();
+    fn new(data_dirs: &[PathBuf], theme_name: &str) -> Self {
+        let mut icons: HashMap<String, HashMap<Bucket, PathBuf>> = HashMap::new();
+
+        // Iterate on all XDG_DATA_DIRS to look for icons.
+        for data_dir in data_dirs {
+            // Find whether each dir has icons, in the only theme we support for now.
+            //
+            // In the future, we might want to parse the index.theme of the theme we want to
+            // load, for instance to handle the proper inheritance hierarchy.
+            let mut icons_dir = data_dir.to_owned();
+            icons_dir.push("icons");
+            icons_dir.push(theme_name);
+
+            for dir_entry in fs::read_dir(icons_dir).unwrap() {
+                let dir_entry = dir_entry.unwrap();
+                let file_name = dir_entry.file_name();
+                let file_name = match file_name.to_str() {
+                    Some(file_name) => file_name,
+                    None => continue,
+                };
```

Comment:
Not sure this should be called "file_name". It's technically a directory we're talking about here.